sycl/doc/design/CompileTimeProperties.md

```diff
@@ -596,6 +620,73 @@ we do not allow non-fundamental types, how do we represent properties like
 allow `std::tuple`, where the type of each element is one of the fundamental
 types listed above.
 
+### IR representation via `@llvm.global.annotations`
+
+Properties that are implemented using
+`[[__sycl_detail__::add_ir_annotations_global_variable()]]`, are represented in 
+LLVM IR as elements in the global variable `@llvm.global.annotations`.
+`@llvm.global.annotations` is an array of `{ ptr, ptr, ptr, i32, ptr}` structs,
+where each struct represents an annotation on a global pointer. The values of
+each field of the struct are as follows:
+
+* pointer to global object to be annotated, global variable in our case. If 
+there are nested annotations on the global variable, constant `getelementptr`
+instructions can be used to specify the level at which the annotation is applied.
+* pointer to global string representing the annotation. Each global string is
+a list of `{name:value}` pairs from the
+`[[__sycl_detail__::add_ir_annotations_global_variable()]]` annotation. Property
+values are converted to strings in the same way as described above, except that
+the `nullptr` value is represented as an empty string, `{name}`.
+* pointer to global string representing the path of the source file where the
+annotation was applied.
+* i32 integer representing the line number in the source file where the
+annotation was applied. 
+* ptr to annotation arguments, `null` pointer in our case 
+
+```cpp
+namespace intel = sycl::ext::intel;
+namespace oneapi = sycl::ext::oneapi;
+
+oneapi::device_global<
+  intel::fpga_mem<
+    int[4],
+    decltype(oneapi::properties(intel::word_size<2>))>,
+  decltype(oneapi::properties(oneapi::host_access_none))> dg;
+```
+
+Below is what the LLVM IR will look like for the SYCL code above.
+
+```llvm
+%"fpga_mem" = type { [4 x i32] }
+%"device_global" = type { %"fpga_mem" }
+
+@.str = private unnamed_addr addrspace(1) constant [25 x i8] c"{sycl-resource:DEFAULT}{5884:2}\00"
+@.str.1 = private unnamed_addr addrspace(1) constant [14 x i8] c"example.cpp\00"
+@.str.2 = private unnamed_addr addrspace(1) constant [11 x i8] c"{sycl-word-size:2}\00"
+
+@dg = dso_local local_unnamed_addr addrspace(1) global { %"device_global" } zeroinitializer
+@llvm.global.annotations = appending global 
+  [2 x { ptr addrspace(1), ptr addrspace(1), ptr addrspace(1), i32, ptr addrspace(1) }]
+  [
+    { ptr addrspace(1), ptr addrspace(1), ptr addrspace(1), i32, ptr addrspace(1) }
+    {
+       ptr addrspace(1) @dg,
+       ptr addrspace(1) @.str.2,
+       ptr addrspace(1) @.str.1,
+       i32 24,
+       ptr addrspace(1) null
+    }, 
+    { ptr addrspace(1), ptr addrspace(1), ptr addrspace(1), i32, ptr addrspace(1) }
+    {
+       ptr addrspace(1) getelementptr inbounds (%"device_global", ptr addrspace(1) @dg, i64 0, i64 0),
```

Comment:
I haven't worked with FE CodeGen much. Is this something that we can do easily @AaronBallman @premanandrao?